app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -0,0 +1,132 @@
+import log from 'loglevel';
+import { captureException } from '@sentry/browser';
+import { isEmpty } from 'lodash';
+import {
+  type IntermediaryStateType,
+  MetaMaskStorageStructure,
+} from './base-store';
+import ExtensionStore from './extension-store';
+import ReadOnlyNetworkStore from './read-only-network-store';
+
+/**
+ * The PersistanceManager class serves as a high-level manager for handling
+ * storage-related operations using a local storage system. It provides methods to read
+ * and write state, manage metadata, and handle errors or corruption in the
+ * underlying storage system.
+ *
+ * Key Responsibilities:
+ *
+ * 1. **State Management:**
+ * - Tracks the most recently retrieved state
+ * - reads state from the storage system
+ * - writes updated state to the storage system
+ *
+ * 2. **Metadata Handling:**
+ * - Manages a `metadata` object containing versioning information for the
+ * state tree. The version is used to ensure consistency and proper
+ * handling of migrations.
+ *
+ * 3. **Error Management:**
+ * - Tracks whether data persistence is failing and logs appropriate errors
+ * - Captures exceptions during write operations and reports them using
+ * Sentry
+ *
+ *
+ * Usage:
+ * The `PersistanceManager` is instantiated with a `localStore`, which is an
+ * implementation of the `BaseStore` class (either `ExtensionStore` or
+ * `ReadOnlyNetworkStore`). It provides methods for setting and retrieving
+ * state, managing metadata, and handling cleanup tasks.
+ */
+export class PersistanceManager {
+  /**
+   * dataPersistenceFailing is a boolean that is set to true if the storage
+   * system attempts to write state and the write operation fails. This is only
+   * used as a way of deduplicating error reports sent to sentry as it is
+   * likely that multiple writes will fail concurrently.
+   */
+  #dataPersistenceFailing: boolean;
+
+  /**
+   * mostRecentRetrievedState is a property that holds the most recent state
+   * successfully retrieved from memory. Due to the nature of async read
+   * operations it is beneficial to have a near real-time snapshot of the state
+   * for sending data to sentry as well as other developer tooling.
+   */
+  #mostRecentRetrievedState: MetaMaskStorageStructure | null;
+
+  /**
+   * metadata is a property that holds the current metadata object. This object
+   * includes a single key which is 'version' and contains the current version
+   * number of the state tree.
+   */
+  metadata?: { version: number };
+
+  #isExtensionInitialized: boolean;
+
+  #localStore: ExtensionStore | ReadOnlyNetworkStore;
+
+  constructor({
+    localStore,
+  }: {
+    localStore: ExtensionStore | ReadOnlyNetworkStore;
+  }) {
+    this.#dataPersistenceFailing = false;
```

Comment:
Nit: We can assign default values where the instance variables are declared, no need for these to be in the constructor

Reply:
addressed in 51cb8dddef